### PR TITLE
Fix nodedevices detach/reattach issues

### PR DESCRIPTION
**Pull request by sannyshao**
(1) Get the device_name from the host on testing not from the cfg
(2) Check the driver with vfio or pci-stub according to libvirt version

Signed-off-by: Jingjing Shao <jishao@redhat.com>

### CODE REVIEW

File: libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py [hunk 1]
```diff
@@ -51,11 +51,19 @@ def do_nodedev_detach_reattach(device_name, params, options=""):
     # check the driver.
     driver = driver_readlink(device_name)
     logging.debug('Driver after detach is %s.', driver)
-    if (driver is None) or (not driver.endswith('pci-stub')):
-        raise error.TestFail("Driver for %s is not pci-stub "
-                             "after nodedev-detach" % (device_name))
+    if libvirt_version.version_compare(1, 1, 1):
```
Comment (chunfuwen) [inline libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: Check code between line 54 to 66:
1)driver is None should raise error regardless libvirt version
2)if not driver.endwith('xx') and 'xx' depends on libvirt version
Therefore maybe those codes can be simplified as below:

if libvirt(1111):
    device_driver_name = 'vfio-ci'
else:
   device_driver_name= 'pci-stub'

if (driver is None) or (not driver.endswith(device_driver_name)):
         raise error.(xx)

Comment (sannyshao) [inline libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: Simplified the code as below, done.

File: libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py [hunk 2]
```diff
@@ -1,6 +1,5 @@
 import os
 import logging
-
 from autotest.client.shared import error
```
Comment (chunfuwen) [inline libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: removed deprecated module

File: libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py [hunk 3]
```diff
@@ -51,27 +50,47 @@ def do_nodedev_detach_reattach(device_name, params, options=""):
     # check the driver.
     driver = driver_readlink(device_name)
     logging.debug('Driver after detach is %s.', driver)
-    if (driver is None) or (not driver.endswith('pci-stub')):
-        raise error.TestFail("Driver for %s is not pci-stub "
```
Comment (chunfuwen) [inline libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: use test.error()instead

File: libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py [hunk 4]
```diff
@@ -51,27 +50,47 @@ def do_nodedev_detach_reattach(device_name, params, options=""):
...
-                             "after nodedev-detach" % (device_name))
+    if libvirt_version.version_compare(1, 1, 1):
+        device_driver_name = 'vfio-pci'
+    else:
+        device_driver_name = 'pci-stub'
+    if (driver is None) or (not driver.endswith(device_driver_name)):
+        raise error.TestFail("Driver for %s is not %s "
```
Comment (chunfuwen) [inline libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: use test.error()instead

File: libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py [hunk 5]
```diff
@@ -51,27 +50,47 @@ def do_nodedev_detach_reattach(device_name, params, options=""):
...
     else:
         pass
-    logging.debug('Nodedev-detach %s successed.', device_name)
+        logging.debug('Nodedev-detach %s successed.', device_name)
     # do the reattach.
     CmdResult = virsh.nodedev_reattach(device_name, options)
     # check the exit_status.
     if CmdResult.exit_status:
-        raise error.TestFail("Failed to reattach %s.\n"
+        raise error.TestFail("Failed to reattach %s\n"
```
Comment (chunfuwen) [inline libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: same with before

File: libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py [hunk 6]
```diff
@@ -51,27 +50,47 @@ def do_nodedev_detach_reattach(device_name, params, options=""):
...
+        device_driver_name = 'pci-stub'
+
+    if (driver is None) or (not driver.endswith(device_driver_name)):
         pass
+        logging.debug('Nodedev-reattach %s successed.', device_name)
     else:
-        raise error.TestFail("Driver for %s is not be reset after "
-                             "nodedev-reattach" % (device_name))
-    logging.debug('Nodedev-reattach %s successed.', device_name)
+        raise error.TestFail("Driver for %s is not %s "
```
Comment (chunfuwen) [inline libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: use test.fail instead

File: libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py (hunk 5 above)
Comment (sannyshao) [reply libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: done

File: libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py (hunk 6 above)
Comment (sannyshao) [reply libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: done

File: libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py (hunk 4 above)
Comment (sannyshao) [reply libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: done

File: libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py (hunk 3 above)
Comment (sannyshao) [reply libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: done

File: libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py (hunk 2 above)
Comment (sannyshao) [reply libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: done

File: libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py [hunk 7]
```diff
@@ -82,29 +13,114 @@ def run(test, params, env):
     (2).Check variables.
     (3).do nodedev_detach_reattach.
```
[the remark this answers is not in the document]
Comment (sannyshao) [reply libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: done

File: libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py [hunk 8]
```diff
@@ -82,29 +13,114 @@ def run(test, params, env):
...
     """
+    def driver_readlink(device_name):
```
[the remark this answers is not in the document]
Comment (sannyshao) [reply libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: done

File: libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py [hunk 9]
```diff
@@ -82,29 +13,114 @@ def run(test, params, env):
...
+        """
+        nodedevxml = nodedev_xml.NodedevXML.new_from_dumpxml(device_name)
+        driver_path = ('%s/driver') % (nodedevxml.get_sysfs_path())
+        try:
+            driver = os.readlink(driver_path)
+        except (OSError, UnicodeError):
+            return None
+        return driver
+
+    def do_nodedev_detach_reattach(device_name, params, options=""):
```
[the remark this answers is not in the document]
Comment (sannyshao) [reply libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: done

File: libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py [hunk 10]
```diff
@@ -82,29 +13,114 @@ def run(test, params, env):
...
+            if unprivileged_user.count('EXAMPLE'):
+                unprivileged_user = 'testacl'
+
+        # do the detach
+        logging.debug('Node device name is %s.', device_name)
+        CmdResult = virsh.nodedev_detach(device_name, options,
+                                         unprivileged_user=unprivileged_user,
+                                         uri=uri)
+        # check the exit_status.
+        if CmdResult.exit_status:
```
[the remark this answers is not in the document]
Comment (sannyshao) [reply libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: done

File: libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py [hunk 11]
```diff
@@ -82,29 +13,114 @@ def run(test, params, env):
...
+        driver = driver_readlink(device_name)
+        logging.debug('Driver after detach is %s.', driver)
+        if libvirt_version.version_compare(1, 1, 1):
+            device_driver_name = 'vfio-pci'
+        else:
+            device_driver_name = 'pci-stub'
+        if (driver is None) or (not driver.endswith(device_driver_name)):
+            test.error("Driver for %s is not %s "
+                       "after nodedev-detach" % (device_name, device_driver_name))
+        else:
```
[the remark this answers is not in the document]
Comment (sannyshao) [reply libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: done

File: libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py [hunk 12]
```diff
@@ -82,29 +13,114 @@ def run(test, params, env):
...
+                       "Detail: %s."
+                       % (device_name, CmdResult.stderr))
+        # check the driver.
+        driver = driver_readlink(device_name)
+        if libvirt_version.version_compare(1, 1, 1):
+            device_driver_name = 'vfio-pci'
+        else:
+            device_driver_name = 'pci-stub'
+
+        if (driver is None) or (not driver.endswith(device_driver_name)):
```
[the remark this answers is not in the document]
Comment (sannyshao) [reply libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: done

**Comment by sannyshao** [comment]
@kylazhang  can you help to check this pr ? thank you in advance.

File: libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py [hunk 13]
```diff
@@ -1,110 +1,115 @@
...
-    """
-    Test virsh nodedev-detach and virsh nodedev-reattach
+    def pci_device_address():
+        """
+        Get the address of pci device
+        """
+        pci_list = virsh.nodedev_list(tree='', cap='pci')
+        pci_devices_address = pci_list.stdout.strip().splitlines()
+        pci_device_address = pci_devices_address[0]
+        return pci_device_address
```
Comment (kylazhang) [inline libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: return pci_devices_address[0] is enough?

File: libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py [hunk 14]
```diff
@@ -1,110 +1,115 @@
...
-    (1).Init variables for test.
-    (2).Check variables.
-    (3).do nodedev_detach_reattach.
-    """
     # Init variables
-    device_name = params.get('nodedev_device_name', 'ENTER.YOUR.PCI.DEVICE')
+    device_name = params.get('nodedev_device_name', 'ENTER.YOUR.PCI.DEVICE.TO.DETACH')
+    if device_name.find('ENTER.YOUR.PCI.DEVICE.TO.DETACH') != -1:
+        replace_name = pci_device_address()
+        device_name = device_name.replace('ENTER.YOUR.PCI.DEVICE.TO.DETACH', replace_name)
```
Comment (kylazhang) [inline libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: so device_name = pci_device_address() ?

File: libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py (hunk 13 above)
Comment (sannyshao) [reply libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: Yes, your concern is right. I realized this when I update these code. Usually I do the detach-reattachment with network pci-device in manual test.  But if so, this case will need a muti-NIC system and need to be separated from other virsh command test run.  I consider the detach-reattachment about NIC pci device will be covered in SRIOV testing. so just choose  pci_devices_address[0]  to do the detach-reattachment in nodedevice test part.

File: libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py [hunk 15]
```diff
@@ -1,110 +1,115 @@
...
+            test.cancel("API acl test not supported in current"
+                        " libvirt version.")
 
-    # do nodedev_detach_reattach
+    # Do nodedev_detach_reattach
     try:
-        do_nodedev_detach_reattach(device_name, params, device_opt)
-    except error.TestFail, e:
+        detach_reattach_nodedev(device_name, params, device_opt)
+    except Exception, e:
```
[the remark this answers is not in the document]
Comment (sannyshao) [reply libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: Yes, I  do the judgement and put out the test fail as expected.
"
  if status_error:
            return
        else:
            test.fail("Test failed in positive case."
                      "error: %s" % e)
"
en
I also do the test before update the code. Below is the test  which output the testfail when prepare something wrong in in positive case.

avocado run --vt-type libvirt  virsh.nodedev_detach_reattach.positive.normal_test
JOB ID     : 4687d351412c974783506f9ee51830e2c157c4f8
JOB LOG    : /root/avocado/job-results/job-2017-11-03T07.08-4687d35/job.log
 (1/1) type_specific.io-github-autotest-libvirt.virsh.nodedev_detach_reattach.positive.normal_test: FAIL (1.71 s)
RESULTS    : PASS 0 | ERROR 0 | FAIL 1 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 5.34 s

Get the error in log as below.
FAIL 1-type_specific.io-github-autotest-libvirt.virsh.nodedev_detach_reattach.positive.normal_test -> TestFail: Test failed in positive case.error: error: Could not find matching device 'ENTER.YOUR.PCI.DEVICE.TO.DETACH'
error: Node device not found: no node device with matching name 'ENTER.YOUR.PCI.DEVICE.TO.DETACH'

File: libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py (hunk 14 above)
Comment (sannyshao) [reply libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: Done，unified to device_address.

File: libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py (hunk 13 above)
Comment (sannyshao) [reply libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: hi kyla, I have update this part to detach-reattach with network pci-device . 
This part  need the muti-NIC server in CI job, I have send the request to kairui and guannan.

**Comment by sannyshao** [comment]
@kylazhang  can you help to review again for the update ? thank you in advance.

File: libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (sannyshao) [reply libvirt/tests/src/virsh_cmd/nodedev/virsh_nodedev_detach_reattach.py]: updated this part to raise exceptions. testfail  not all the exceptions

**Pull request merged**